### ISSUE TITLE
Project doesn't run on Windows 10 64-bit (possibly related to node path?)

### ISSUE THREAD

**Issue by alexandrethsilva** [patternplate/patternplate#257]
@marionebl First of all, thanks a lot for this project. The concepts behind it are really interesting and it has been a great experience to use it on OS X.

Regarding the issue, when running the latest version (`2.5.25`) I had a project break when being used by a developer trying to run it on a Windows 10 64-bit machine and, upon investigation, I had a feeling that it was in one case (my project) a `path` related issue, and on another (the getting-started sample) something else I didn't yet figure.

Replicating it even with the example project from the "getting started" guide is as easy as running `yarn start` on a fresh install and trying to open the docs on `http://localhost:1337`.

I've seen that at least a couple PRs have been made in attempts to solve Windows-related path handling issues and am assuming that currently the project is being mainly developed on UNIX-based machines, right?

Do you guys have an idea on how feasible it is to correct these issues in the short-term? I develop on OS X myself and since I don't usually have to offer cross-platform projects I find myself a bit unsure on how reliable it is to use patternplate for such projects?

I'm currently investigating the failure points and could offer a PR from my fork in the future, but just want to avoid making double-work in case you're on it already.

**Btw, this is the current error being thrown on the getting-started example:**

```bash
Z:\Development\Playgrounds\my-patternplate>yarn start
yarn run v1.9.4
$ patternplate start
√ Started on http://localhost:1337
events.js:167
      throw er; // Unhandled 'error' event
      ^

Error: write EPIPE
    at ChildProcess.target._send (internal/child_process.js:742:20)
    at ChildProcess.target.send (internal/child_process.js:626:19)
    at send (Z:\Development\Playgrounds\my-patternplate\node_modules\@patternplate\api\lib\create-compiler.js:328:14)
    at Timeout.setInterval [as _onTimeout] (Z:\Development\Playgrounds\my-patternplate\node_modules\@patternplate\api\lib\create-compiler.js:372:23)
    at ontimeout (timers.js:425:11)
    at tryOnTimeout (timers.js:289:5)
    at listOnTimeout (timers.js:252:5)
    at Timer.processTimers (timers.js:212:10)
Emitted 'error' event at:
    at process.nextTick (internal/child_process.js:746:39)
    at process._tickCallback (internal/process/next_tick.js:61:11)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

```

**The compiled file crashing on `node_modules`:**
```javascript
module.exports.createCompiler = function createCompiler({
  config,
  cwd,
  target = ""
}) {
  return new Promise(function ($return, $error) {
    let worker;

    const send = payload => {
      if (!worker || !worker.send || !worker.connected) {
        return;
      }

      worker.send(ARSON.stringify(payload)); // <------ LINE 328
    };

    let watching = false;
    const queue = [];
    let listeners = [];

    const next = message => listeners.forEach(listener => listener.next(message));

    const start = () => {
      const workerPath = path.join(__dirname, "compiler-worker.js");
      debug(`starting compiler worker at ${workerPath}`);
      const cp = fork(workerPath, dargs({
        cwd,
        target,
        config: ARSON.stringify(config)
      }), OPTS);
      let stderr = ``;
      let stdout = ``;
      cp.stdout.on("data", data => {
        stdout += String(data);
        debug(`stdout ${workerPath}: ${data}`);
      });
      cp.stderr.on("data", data => {
        stderr += String(data);
        debug(`stderr ${workerPath}: ${data}`);
      });
      cp.once("close", code => {
        if (code !== 0) {
          queue.unshift({
            type: "exception",
            payload: {
              code,
              stdout,
              stderr: [`Could not start compiler worker for "${target}"`, stderr].join("\n")
            }
          });
          next(queue);
        }
      });
      return cp;
    };

    worker = start();
    setInterval(() => send({ // <----- LINE 372
      type: "heartbeat"
    }), 500);

    const onMessage = envelope => {
      const _ARSON$parse = ARSON.parse(envelope),
            type = _ARSON$parse.type,
            target = _ARSON$parse.target,
            payload = _ARSON$parse.payload;

      switch (type) {
        case "ready":
          return send({
            type: "start"
          });

        case "done":
          {
            debug({
              type,
              target
            });
            const fs = new MemoryFilesystem(payload);
            queue.unshift({
              type,
              target,
              payload: {
                fs
              }
            });
            return next(queue);
          }

        case "start":
          debug({
            type,
            target
          });
          queue.unshift({
            type,
            target,
            payload
          });
          return next(queue);

        case "error":
          if (Array.isArray(payload)) {
            return payload.forEach(p => console.error(p.message));
          }

          return console.error(payload.message);

        case "shutdown":
          debug({
            type,
            target
          });
          send({
            type: "stop"
          });
          worker = start();
          worker.on("message", onMessage);

          if (watching) {
            send({
              type: "watch"
            });
          }

      }
    };

    worker.on("message", onMessage);
    const observable = new Observable(observer => {
      if (!watching) {
        watching = true;
        send({
          type: "watch"
        });
      }

      listeners.push(observer);
      return () => {
        listeners = listeners.filter(listener => listener !== observer);
      };
    });
    observable.queue = queue;

    observable.stop = () => {
      send({
        type: "stop"
      });
    };

    return $return(observable);
  }.$asyncbind(this));
};
```

Do you happen to have any insights?

**Comment by marionebl** [comment]
Thanks for the report @alexandrethsilva. (Glad you like working with patternplate 🎉 )

## TL;DR

I'll investigate this particular issue with the getting started guide over the weekend. Looks like there is some exception being thrown when patternplate tries to start the internal webpack workers. 

That being said I'd appreciate a PR addressing issues where you found them. 👍 

Any chance you could also provide error details for the other project you mentioned?

## Windows support 

I def. want to support win32 systems and invest the additional work required to do so. Your guess about me using OSX for development is right, though, thus the odd `path` problem cropping up here and there in the project (in its current state). 

I'm in the process of improving patterplate's overall sturdiness by 

* porting all packages to TypeScript
* introducing unit/integration tests for all packages

This will involve a basic AppVeyor test down the line, which should catch this type of problem in the future.

**Comment by alexandrethsilva** [comment]
Thanks for the quick Feedback @marionebl, I'll be glad to open a PR as soon as I can identify what's going on. 👌 

I'll rerun my project against `ndb` and update this issue with the output being thrown there right after.

**Comment by alexandrethsilva** [comment]
So, in my Project specifically, it's throwing also on `create-compiler`:
```sh
events.js:167 Uncaught Error: write EPIPE
    at ChildProcess.target._send (internal/child_process.js:742:20)
    at ChildProcess.target.send (internal/child_process.js:626:19)
    at send (Z:\Development\Playgrounds\my-project\node_modules\@patternplate\api\lib\create-compiler.js:328:14)
    at Timeout.setInterval [as _onTimeout] (Z:\Development\Playgrounds\my-project\node_modules\@patternplate\api\lib\create-compiler.js:372:23)
    at ontimeout (timers.js:425:11)
    at tryOnTimeout (timers.js:289:5)
    at listOnTimeout (timers.js:252:5)
    at Timer.processTimers (timers.js:212:10)
Emitted 'error' event at:
    at process.nextTick (internal/child_process.js:746:39)
    at process._tickCallback (internal/process/next_tick.js:61:11)
```

Here's the piece:


```js
const onMessage = envelope => {
      const {type, target, payload} = ARSON.parse(envelope);

      switch (type) {
        case "ready": {
          return send({type: "start"});
        }
        case "done": {
          debug({type, target});
          const fs = new MemoryFilesystem(payload);
          queue.unshift({type, target, payload: {fs}});
          return next(queue);
        }
        case "start": {
          debug({type, target});
          queue.unshift({type, target, payload});
          return next(queue);
        }
        case "error": {
          if (Array.isArray(payload)) {
            return payload.forEach(p => console.error(p.message))
          }
          return console.error(payload.message); // <---- CRASHES HERE @LINE 90
        }
        case "shutdown": {
          debug({type, target});
          send({type: "stop"});

          worker = start();

          worker.on("message", onMessage);

          if (watching) {
            send({type: "watch"});
          }
        }
    }
  };
```

So it tries to pull the `message` off the `payload` object, but the key doesn't exist. The shape of it is actually the following:

```js
{
  envelope: String,
  payload: {
    dependencies: [Object],
    details: String, // <-- value: Identical to `error.details`
    error: {
      details: String,
      missing: [ // <-- Apparently a list of the fallbacks it attempted to resolve
        "Z:\Development\Playgrounds\my-project\libadgesdemo.css",
        "Z:\Development\Playgrounds\my-project\libadgesdemo.css.wasm",
        "Z:\Development\Playgrounds\my-project\libadgesdemo.css.mjs",
        "Z:\Development\Playgrounds\my-project\libadgesdemo.css.js",
        "Z:\Development\Playgrounds\my-project\libadgesdemo.css.json"
      ]
    },
    missing: String,   // <-- value: Identical to `error.missing`,
    module: Object,
    name: String,      // <-- value: "ModuleNotFoundError",
    origin: Object,    // <-- value: Identical to the `module` key
    target: String     // <-- value: "node",
    this: undefined,
    type: String,       // <-- value: "error"
    _ARSON$parse: Object
  }
}
```

And what I meant with my suspicion that it's related to `path` Resolution is that this:
`"Z:\Development\Playgrounds\my-project\libadgesdemo.css"`

Should be this:
`"Z:\Development\Playgrounds\my-project\lib\badges\demo.css"`

So it seems to be doing something wrong there, since it's throwing this for each one of the file references inside the `lib` folder.

**Comment by alexandrethsilva** [comment]
@marionebl Great stuff! Thanks a lot 🚀

**Comment by marionebl** [comment]
> @marionebl Great stuff! Thanks a lot 🚀
> Just a side question to it: it's not released on npm yet, right?

In the process of publishing: 😁 

<img width="1046" alt="screen shot 2018-09-17 at 16 34 17" src="https://user-images.githubusercontent.com/4248851/45629846-a2dcaf80-ba97-11e8-8a05-a286a32a11a0.png">

**Comment by marionebl** [comment]
Released via `3.0.1` 🎉
